torch/csrc/distributed/c10d/FlightRecorder.cpp

```diff
@@ -20,6 +20,403 @@
 
 namespace c10d {
 
+template <typename T, typename EntryT, typename EventType>
+std::optional<size_t> recordImpl(
+    T& obj,
```

Comment:
what is 'obj'?  is this supposed to represent a ProcessGroup or something?

I wonder if we need to use Templating for T and EntryT, or we should just use templating for EventType and use polymorphism for the rest? thoughts?

Reply:
No it is a flight recorder instance so that we can differentiate between different situations (cudaEvent vs c10::Event). The signature of this function requires a Event. Actually in the next PR, it does not need this EventType.

Reply:
Also we need obj to access the members of the struct. At least for this case. The only two API we can leverage polymorphism is get_dump_json and get_dump, which I think make it templated is better? IDK since the function being called is also template not member function, so I just make them all templated.

Reply:
can we templatize this a bit tighter? I.e. have this take a FlightRecorder<T> instead of just T?

Reply:
Similar question to why we'd need `T` and `EntryT` --
if we previously defined `class Entry`, would it just work if we extend it to `class Entry<EventType>`?